hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/H2LegacyDialect.java

```diff
@@ -0,0 +1,707 @@
+
+		functionFactory.pi();
+		functionFactory.cot();
+		functionFactory.radians();
+		functionFactory.degrees();
+		functionFactory.log10();
+		functionFactory.rand();
+		functionFactory.truncate();
+		functionFactory.soundex();
+		functionFactory.translate();
+		functionFactory.bitand();
+		functionFactory.bitor();
+		functionFactory.bitxor();
+		functionFactory.bitAndOr();
+		functionFactory.yearMonthDay();
+		functionFactory.hourMinuteSecond();
+		functionFactory.dayOfWeekMonthYear();
+		functionFactory.weekQuarter();
+		functionFactory.daynameMonthname();
+		if ( useLocalTime ) {
+			functionFactory.localtimeLocaltimestamp();
+		}
+		functionFactory.bitLength();
+		functionFactory.octetLength();
+		functionFactory.ascii();
+		functionFactory.octetLength();
+		functionFactory.space();
+		functionFactory.repeat();
+		functionFactory.chr_char();
+		functionFactory.instr();
+		functionFactory.substr();
+		//also natively supports ANSI-style substring()
+		functionFactory.position();
+		functionFactory.trim1();
+		functionFactory.concat_pipeOperator();
+		functionFactory.nowCurdateCurtime();
+		functionFactory.sysdate();
+		functionFactory.insert();
+//		functionFactory.everyAny(); //this would work too
+		functionFactory.everyAny_boolAndOr();
+		functionFactory.median();
+		functionFactory.stddevPopSamp();
+		functionFactory.varPopSamp();
+		if ( getVersion().isSame( 1, 4, 200 ) ) {
+			// See https://github.com/h2database/h2database/issues/2518
+			functionFactory.format_toChar();
+		}
+		else {
+			functionFactory.format_formatdatetime();
+		}
+		functionFactory.rownum();
+		if ( getVersion().isSameOrAfter( 1, 4, 200 ) ) {
+			functionFactory.windowFunctions();
+			functionFactory.listagg( null );
+			if ( getVersion().isSameOrAfter( 2 ) ) {
+				functionFactory.inverseDistributionOrderedSetAggregates();
+				functionFactory.hypotheticalOrderedSetAggregates();
+			}
+		}
+		else {
+			functionFactory.listagg_groupConcat();
+		}
+	}
+
+	@Override
+	public void augmentPhysicalTableTypes(List<String> tableTypesList) {
+		if ( getVersion().isSameOrAfter( 2 ) ) {
+			tableTypesList.add( "BASE TABLE" );
+		}
+	}
+
+	@Override
+	protected Integer resolveSqlTypeCode(String columnTypeName, TypeConfiguration typeConfiguration) {
+		switch ( columnTypeName ) {
+			case "FLOAT(24)":
+				// Use REAL instead of FLOAT to get Float as recommended Java type
+				return Types.REAL;
+		}
+		return super.resolveSqlTypeCode( columnTypeName, typeConfiguration );
+	}
+
+	@Override
+	public JdbcType resolveSqlTypeDescriptor(
+			String columnTypeName,
+			int jdbcTypeCode,
+			int precision,
+			int scale,
+			JdbcTypeRegistry jdbcTypeRegistry) {
+		// As of H2 2.0 we get a FLOAT type code even though it is a DOUBLE
+		if ( jdbcTypeCode == FLOAT && "DOUBLE PRECISION".equals( columnTypeName ) ) {
+			return jdbcTypeRegistry.getDescriptor( DOUBLE );
+		}
+		return super.resolveSqlTypeDescriptor( columnTypeName, jdbcTypeCode, precision, scale, jdbcTypeRegistry );
```

Comment:
`REAL` is reported as `Types.REAL`.
`DOUBLE PRECISION` is reported as `Types.DOUBLE`.

`FLOAT` is reported as `Types.DOUBLE` by H2 1.4.196 and older versions.
`FLOAT` is reported as either `Types.REAL` or `Types.DOUBLE` by H2 1.4.197–1.4.200 depending on precision.
`FLOAT` is reported as `Types.FLOAT` by H2 2.0.202 and newer versions, but actual type is either `REAL` or `DOUBLE PRECISION` anyway depending on precision.

Reply:
If you have ideas to improve the dialect, please create an issue for that and provide a PR